src/Provider/AbstractProvider.php

```diff
@@ -135,17 +176,46 @@ public function setScopes(array $scopes)
         $this->scopes = $scopes;
     }
 
+    /**
+     * Get a new random string to use for auth state.
+     *
+     * @param  integer $length
+     * @return string
+     */
+    protected function getRandomState($length = 32)
+    {
+        $generator = $this
+            ->getRandomFactory()
+            ->getMediumStrengthGenerator();
+
+        return $generator->generateString($length);
+    }
+
     public function getAuthorizationUrl(array $options = [])
     {
-        $this->state = isset($options['state']) ? $options['state'] : md5(uniqid(rand(), true));
+        if (empty($options['state'])) {
+            $options['state'] = $this->getRandomState();
+        }
+
+        // Store the state, it may need to be accessed later.
+        $this->state = $options['state'];
+
+        $options += [
+            // Do not set the default state here! The random generator takes a
+            // non-trivial amount of time to run.
+            'response_type'   => 'code',
+            'approval_prompt' => 'auto',
+        ];
+
+        $scopes = is_array($this->scopes) ? implode($this->scopeSeparator, $this->scopes) : $this->scopes;
```

Comment:
Would it make sense to allow the `scopes` to be injected with `$options`? Seems like it could be a useful thing, but obviously out of scope with this PR.